### ISSUE TITLE
Llama4 FP8 + EP: Dynamo recompile limit exceeded during SAC backward

### ISSUE THREAD

**Issue by acisseJZhong** [pytorch/torchtitan#2958]
## Bug

Llama4 debug model with FP8 quantization and expert parallelism (EP=2) crashes during backward pass with `FailOnRecompileLimitHit: Hard failure due to fullgraph=True`.

## Repro

```bash
NCCL_NVLS_ENABLE=0 MODULE=llama4 CONFIG=llama4_debugmodel_fp8 NGPU=4 ./run_train.sh \
  --parallelism.pipeline_parallel_degree 1 \
  --parallelism.data_parallel_shard_degree 4 \
  --parallelism.tensor_parallel_degree 1 \
  --parallelism.expert_parallel_degree 2 \
  --parallelism.expert_tensor_parallel_degree 1
```

## Error

```
torch._dynamo.exc.FailOnRecompileLimitHit: Hard failure due to fullgraph=True
```

Dynamo hits the recompile limit (8) on `WeakValueDictionary.__setitem__` inside the checkpoint `pack_hook` during activation checkpointing recomputation in backward.

## Traceback (condensed)

```
loss.backward()
  → scatter_add.py:31 _backward: ctx.saved_tensors
    → checkpoint.py:1161 unpack_hook → recompute_fn
      → llama4/model.py:102: self.moe(self.ffn_norm(h))
        → rms_norm → checkpoint.py:1104 pack_hook
          → WeakValueDictionary.__setitem__
            → Dynamo recompile limit exceeded (8)
```

## Configuration

- Model: Llama4 debugmodel FP8
- Parallelism: FSDP=4, EP=2, TP=1, PP=1, ETP=1
- Selective activation checkpointing + torch.compile (fullgraph=True)

## Notes

The interaction between the custom `scatter_add` op's backward, selective activation checkpointing recomputation, and torch.compile's `fullgraph=True` causes excessive Dynamo recompilations of `weakref.WeakValueDictionary.__setitem__` in the checkpoint pack hook.

**Comment by xmfan** [comment]
@anijain2305 feel free to retriage. So the usual workaround would be to disable the pack hook, but we can't do that with fullgraph=True (and the AC problem). Marked as high pri because I suspect this affects every MoE.

This looks unrelated to the latest scatter add custom op, probably related to fp8. The failure chain is:

  1. Forward pass (compiled with fullgraph=True): The Llama4TransformerBlock runs through SAC (_checkpoint_hook). During forward, the pack_hook saves tensors via frame.weak_holders.append(weakref.ref(holder)). This
  is traced by Dynamo (no @torch._dynamo.disable on _checkpoint_hook.pack_hook, unlike _recomputation_hook.pack_hook which does have it).
  2. Backward pass: loss.backward() triggers unpack_hook → recompute_fn re-runs the forward. During recomputation, the nested _recomputation_hook.pack_hook (which IS dynamo-disabled) captures tensors. But this
  re-invokes the compiled sub-modules of the MoE block (router, reorderer, etc. — all compiled with fullgraph=True in apply_compile_sparse).
  3. Recompile trigger: With EP enabled, the MoE code has dynamic token counts (mark_dynamic(x, 0) in compile.py:141). With FP8, the ScaledGroupedMMTensor from torchao adds quantization scaling factor state. During
  each backward recomputation, Dynamo sees:
    - Different weak_holders list lengths — the list grows with each pack_hook call, and Dynamo guards on collection structure
    - Different dynamic shapes from EP — num_tokens_per_expert varies across steps
    - FP8 scaling factor state — Float8TrainingOpConfig uses dynamic per-tensor scaling that changes each iteration
  4. fullgraph=True makes it fatal: Normally Dynamo would insert a graph break and fall back to eager. But fullgraph=True (compile.py:103) converts graph breaks into hard errors. After 8 recompilations on the
  WeakValueDictionary.__setitem__ guards, Dynamo raises FailOnRecompileLimitHit.

  Why FP8 + EP specifically

  - Without EP: Token counts are static per expert, so _run_experts_grouped_mm doesn't need mark_dynamic. Fewer guard invalidations.
  - Without FP8: No ScaledGroupedMMTensor subclass, so fewer dynamic internal states for Dynamo to guard on.
  - With both: The combination of dynamic token shapes AND dynamic FP8 scaling factors creates enough guard variation that the recompile limit (8) is exceeded, specifically during the SAC recompute path where the
  _checkpoint_hook.pack_hook's weak_holders list state also varies.

**Comment by anijain2305** [comment]
I am getting this error right now - https://www.internalfb.com/intern/paste/P2291486534/

Is there any fix that I should take? I will keep on looking, just wondering if I am missing something.

**Comment by xmfan** [comment]
This is being masked by another error now cc @laithsakka 

From Claude:
```
Where u22 (the unbacked symbol) comes from                                                                                                                                                   
                                               
  You enabled expert_parallel_degree=2. In an EP MoE, the local input to each expert's grouped GEMM has a data-dependent number of tokens (depends on routing → all-to-all dispatch). So A     
  arriving at _Float8GroupedMM is (M, K) with M unbacked.                                                                                              
                                                                                                                                                                                               
  Inside backward, the problematic chain is:                                                                                                                                                   
                                               
  # fp8_grouped_mm.py:286                                                                                                                                                                      
  grad_out_data_colwise, ... = triton_fp8_per_group_colwise_scales_dual(                                                                               
      padded_grad_output, padded_A, padded_group_end_offsets, ...                                                                                                                              
  )
  # shape (M, N), column-major, strides=(1, M)                                                                                                                                                 
                                                                                                                                                                                               
  # line 295                                   
  grad_output_t_data_row_major = grad_out_data_colwise.t()                                                                                                                                     
  # shape (N, M), strides=(M, 1)   ← stride(-2) is now the unbacked M                                                                                  
                                                                                                                                                                                               
  # line 300  ← FAILS                                                                                                                                                                          
  assert not _is_column_major(grad_output_t_data_row_major)                                                                                                                                    
  #   _is_column_major := stride(-2) == 1                                                                                                                                                      
  #   evaluates Eq(M, 1) where M is unbacked → GuardOnDataDependentSymNode      
```

**Comment by laithsakka** [comment]
assert not _is_column_major(grad_output_t_data_row_major)                                                                                                                                    
  in 
    File "/home/anijain/.conda/envs/pytorch2/lib/python3.12/site-packages/torchao/prototype/moe_training/fp8_grouped_mm.py", line 303, in backward
      assert not _is_column_major(grad_output_t_data_row_major), (
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
need to be rewritten using a torch._check so it become a runtime assertion 
See 
https://docs.pytorch.org/docs/stable/user_guide/torch_compiler/compile/dynamic_shapes_troubleshooting_guardon_errors.html#torch-check-cond-msg-fn

**Comment by xmfan** [comment]
This is fixed off nightly pytorch, torchtitan main, ao main
